pkg/services/authz/zanzana/store.go

```diff
@@ -1,13 +1,216 @@
+	grafanaDBCfg, zanzanaDBCfg, err := parseConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse database config: %w", err)
+	}
+
+	switch grafanaDBCfg.Type {
+	case migrator.SQLite:
+		return memory.New(), nil
+	case migrator.MySQL:
+		// For mysql we need to pass parseTime parameter in connection string
+		connStr := grafanaDBCfg.ConnectionString + "&parseTime=true"
+		if err := runMigrations(cfg, migrator.MySQL, connStr, assets.MySQLMigrationDir); err != nil {
+			return nil, fmt.Errorf("failed to run migrations: %w", err)
+		}
+		return mysql.New(connStr, zanzanaDBCfg)
+	case migrator.Postgres:
+		if err := runMigrations(cfg, migrator.Postgres, grafanaDBCfg.ConnectionString, assets.PostgresMigrationDir); err != nil {
+			return nil, fmt.Errorf("failed to run migrations: %w", err)
+		}
+		return postgres.New(grafanaDBCfg.ConnectionString, zanzanaDBCfg)
+	}
+
+	// Should never happen
+	return nil, fmt.Errorf("unsupported database engine: %s", grafanaDBCfg.Type)
+}
+
+func NewEmbeddedStore(cfg *setting.Cfg, db db.DB) (storage.OpenFGADatastore, error) {
+	grafanaDBCfg, zanzanaDBCfg, err := parseConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse database config: %w", err)
+	}
+
+	m := migrator.NewMigrator(db.GetEngine(), cfg)
+
+	switch grafanaDBCfg.Type {
+	case migrator.SQLite:
+		// FIXME(kalleep): At the moment sqlite is not a supported data store.
+		// So we just return in memory store for now.
+		return memory.New(), nil
+	case migrator.MySQL:
+		if err := runMigrationsWithMigrator(m, cfg, assets.MySQLMigrationDir); err != nil {
+			return nil, fmt.Errorf("failed to run migrations: %w", err)
+		}
+
+		// For mysql we need to pass parseTime parameter in connection string
+		return mysql.New(grafanaDBCfg.ConnectionString+"&parseTime=true", zanzanaDBCfg)
+	case migrator.Postgres:
+		if err := runMigrationsWithMigrator(m, cfg, assets.PostgresMigrationDir); err != nil {
+			return nil, fmt.Errorf("failed to run migrations: %w", err)
+		}
+
+		return postgres.New(grafanaDBCfg.ConnectionString, zanzanaDBCfg)
+	}
+
+	// Should never happen
+	return nil, fmt.Errorf("unsupported database engine: %s", db.GetDialect().DriverName())
+}
+
+func parseConfig(cfg *setting.Cfg) (*sqlstore.DatabaseConfig, *sqlcommon.Config, error) {
+	sec := cfg.Raw.Section("database")
+	grafanaDBCfg, err := sqlstore.NewDatabaseConfig(cfg, nil)
+	if err != nil {
+		return nil, nil, nil
+	}
+
+	// There are a couple of things that are not supported by
+	zanzanaDBCfg := &sqlcommon.Config{
+		// FIXME(kalleep): set logger when we have written a adapter
+		Logger: logger.NewNoopLogger(),
+		// MaxTuplesPerWriteField: 0,
+		// MaxTypesPerModelField:  0,
+		MaxOpenConns:    grafanaDBCfg.MaxOpenConn,
+		MaxIdleConns:    grafanaDBCfg.MaxIdleConn,
+		ConnMaxLifetime: time.Duration(grafanaDBCfg.ConnMaxLifetime) * time.Second,
+		ExportMetrics:   sec.Key("instrument_queries").MustBool(false),
```

Comment:
I'm not sure, it sounds a bit confusing - `instrument_queries` option enables traces for DB queries, but `ExportMetrics` enables DB metrics in openFGA. Maybe we should have separate option or run it by default? However, there're other openFGA metrics collected, but no config option to enable/disable it. So maybe that's fine if it enables only DB-specific metrics.

Reply:
Yeah so there is a comment around this config option
```
# Set to true to add metrics and tracing for database queries.
instrument_queries = false
```

So I figured that it fits